Woo email customizer - style settings improvements [MAILPOET-4180]

costasovo opened the pull request:
## Description

This PR addresses the issue that, when editing Woo transactional emails with the email editor, some style settings were not applied consistently to both Woo content and content added in the editor.

### Changes

1. It is now possible to set font size for H1, H2, H3
2. It is now possible to set font family and font color for headings and the rest of the content
4. The H1 font size setting is applied to the main heading in the header
5. All headings use branding color. Note: Headings added via Text block cab still have a color set via the TinyMCE
6. Heading font size, font family and font color are also applied to headings in Woo block (order details, new customer, etc)
7. Text font size is also applied to content in Woo block
8. The changes applied to the Woo content are not applied immediately after updating to the new plugin version, but the user needs to save the email template in the Woo email editor. This is done to make sure the user is aware of the changes.

#### Before
![screenshot-2024-07-16-at-10 52 14](https://github.com/user-attachments/assets/4667a129-3c03-4b35-9664-527565f8d1b3)

#### After
<img width="982" alt="Screenshot 2024-07-23 at 12 35 02" src="https://github.com/user-attachments/assets/4670d061-88be-434f-8800-eaefb1f01eaf">



## Code review notes

I recommend going through the commits one by one.

## QA notes

Please test the customizer with various Woo transactional emails. The changes were made in the editor and in the renderer, so we also need to check the emails that were sent.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4180]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes -- I edited two JS files in the current editor, but I chose not to convert those to TS


[MAILPOET-4180]: https://mailpoet.atlassian.net/browse/MAILPOET-4180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Apply style settings on Woo Blocks in the Woo Customizer

The font size was not applied at all, but Woo content comes with inlined font-family
styles so we need to use !important to override it.
[MAILPOET-4180]

costasovo committed:
Use font family from Woo Customizer in Woo email styles

Woo has a stylesheet that is later inlined into the email content by Emogrifier.
In this commit I replace all font family declarations by the font family set
in the Woo customizer.
[MAILPOET-4180]

costasovo committed:
Replaced inlined font-family definitions in email content from Woo

[MAILPOET-4180]

costasovo committed:
Unify headings in Woo Customizer to use branding color and global font

[MAILPOET-4180]

costasovo committed:
Enforce the same font family for headings in the Woo transaction email.

In the Woo email customizer, we have only one font family.
The newsletter data carry different default font families for headings.
In this commit, we enforce that the renderer uses the same font family
for text and headings.
This only applies to the part of the email that is handled by the MailPoet renderer.
The part that comes from the Woo itself is handled differently.
See previous commits.
I used clone to make sure we don't accidentally save the settings.
It wouldn't probably cause any damage because we don't expose these settings to users, but to be safe.
[MAILPOET-4180]

costasovo committed:
Set font size for Woo content

This set font size for the WooContent block.
The font size is applied to elements that don't have an inlined font size.
[MAILPOET-4180]

costasovo committed:
Enforce a special heading color for the Woo email header

[MAILPOET-4180]

costasovo committed:
Add a check to apply style fixes only after the email was re-saved

Some styles newly applied for the Woo Content blocks might
make the blocks look broken (e.g., too big font applied on the order items table).
This commit adds a check for existing emails, and for sites with existing customized the Woo email template
they are not applied unless a user saves the email in the editor.
[MAILPOET-4180]

costasovo committed:
Allow setting font sizes for headings in Woo customizer

[MAILPOET-4180]

costasovo committed:
Apply heading font sizes to Woo content in emails

[MAILPOET-4180]

costasovo committed:
Add Typography label and change label positioning in Styles panel

[MAILPOET-4180]

costasovo committed:
Add heading font family to Woo email customizer

[MAILPOET-4180]

costasovo committed:
Render heading font family in Woo transactional emails

[MAILPOET-4180]

costasovo committed:
Refactor modifications of styles from Woo to an extra method

[MAILPOET-4180]

costasovo committed:
Remove unused CSS for main heading font size in Woo emails

This heading will from now on respect H1 font-size which default
value is 36px.
[MAILPOET-4180]

costasovo committed:
Improve test coverage for Woo emails renderer

[MAILPOET-4180]

NeosinneR commented on mailpoet/lib/WooCommerce/TransactionalEmails/Renderer.php:
```diff
@@ -70,6 +70,13 @@ public function getHTMLAfterContent() {
+    $isSavedWithStyledWooBlock = $newsletter->getGlobalStyle('woocommerce', 'isSavedWithUpdatedStyles');
+    // For Backward compatibility do not apply styles for content unless the template was edited with the editor
+    // and user visually checked and is aware of updated styles feature.
+    // This check is save to remove after 2025-07-31
```
Is there an assumption that everyone will re-save their current Woo templates within a year? As far as I understand, if we remove this check after a year, it can still break the styling of emails that were saved before now but haven't been re-saved since then. Unless there is a need to remove this check, I think it would be safer to keep it so we don't have to debug why something broke in a year's time.

costasovo replied:
> Is there an assumption that everyone will re-save their current Woo templates within a year?

Yep. That's how I was thinking about it. I don't like this type of condition, so I wanted to figure out a way to get rid of it. But you're right. It may cause unexpected issues after a year or so. I'll remove that part of comment.

NeosinneR reviewed (changes requested):
Overall, the changes look good to me, @costasovo, but I'd like to check with you regarding two things.

I only realized this while testing it locally, but a light brand color can make content headings hard (or impossible) to read. The main heading is fine as it adjusts by changing the font color from white to black automatically.

<img width="707" alt="Screenshot 2024-07-22 at 15 05 15" src="https://github.com/user-attachments/assets/b99404c2-5ca0-41bc-ac37-59f024e539e1">

What would you say about allowing color change for all headings? 

**Before:**
<img width="136" alt="Screenshot 2024-07-22 at 15 10 38" src="https://github.com/user-attachments/assets/623be396-9764-49a4-8303-cb698c9b1183">

**After:**
<img width="146" alt="Screenshot 2024-07-22 at 15 11 34" src="https://github.com/user-attachments/assets/ae3a59d3-9066-4bb6-80bd-c97ed267763a">

<img width="682" alt="Screenshot 2024-07-22 at 15 12 50" src="https://github.com/user-attachments/assets/e267a76e-ec80-4084-aa6f-e0310422f6ad">

costasovo commented:
@NeosinneR Good catch about the light branding colors. I'll add the heading color setting.

costasovo committed:
Add heading font color control to Woo email customizer

[MAILPOET-4180]

costasovo committed:
Add support for heading font color to Woo transaction email renderer

[MAILPOET-4180]

costasovo committed:
Remove comments about deleting checks for BC support

We can't be sure that after a year, all customers will re-save the email.
To avoid future issues, we will keep the check in place.
[MAILPOET-4180]

costasovo commented:
@NeosinneR I implemented the proposed changes. The heading colors are, by default, not set, and in this case the headings use the branding colors. I believe that this is what users will need in the majority of cases.

NeosinneR reviewed (approved):
LGTM, thanks for the changes 👍

Aschepikov reviewed (approved):
GJ!
